linera-views/DESIGN.md

```diff
@@ -0,0 +1,140 @@
+* The `ValueSplittingKeyValueStoreClient<K>` implements a client for which the
+size of the values is unbounded, on top of another client for which it is bounded.
+(Some databases have strict limitations on the value size.)
+
+## Views
+
+A view is a container that mimics some of the properties of classic containers such
+as `VecDeque`, `BTreeMap`, etc. with the following goals:
+* The data structure can be written to the key-value store and loaded from there.
+* The changes are first done in memory and accessed there. The changes can be committed
+to the key-value store client when adequate.
+
+There are various views corresponding to several use cases. All of those are
+implemented by mapping the values in their specific types to their serializations.
+
+## Allowed keys, base_key and prefixes
+
+The keys are constructed step by step from the prefix to the full key. So, whenever
+we have a view or any other object then we have a `base_key` associated with it and
+with no other objects. A view is associated to many keys, all of whom share the same
+`base_key` as prefix.
+This leads us to introduce a trait `Context` that can be implemented with a specific
+`KeyValueStoreClient` client and a `base_key`.
+
+Another issue to consider is that we do not want to just store the values, but we
+also need to accommodate other features:
+* Administrative data (like journal) need to store some data starting from the `base_key`.
+* Some of the values need to be split into several smaller values.
+
+The rule for constructing are the following:
+* For the construction of `struct` objects of associated base key `base_key` we do the
+following: If the corresponding type has entries `entry_0`, ..., `entry_k` then the
+base key of the object associated with the `k`-th entry is `[base_key * * * *]` with `[* * * *]`
+the serialization of the number `k` considered as a `u32`.
+* For each view type, we have some associated data. Some are the hash, others the counts
+(e.g. for `LogView` and `QueueView`) and some the index. The corresponding enum is
+named `KeyTag` for each of those containers. For the index, the values are serialized.
+* For a given base key `base_key` the keys of the form `[base_key tag]` with tags in
+`0..MIN_VIEW_TAG` are reserved for administrative data.
+
+These constructions are done in order to satisfy the following property: The set of keys
+of a view is prefix-free. That is if a key is of the form `key = [a b]` then `a` is
+not a key.
+
+
+# Chosen implementation
+
+## API of the `KeyValueStoreClient`.
+
+The design of the key-value store client is done in the following way:
+* We can put a `(key, value)` and delete a `key`.
+* We can delete keys having a specific prefix.
+* We can retrieve the list of keys having a specific prefix. Similarly, we can
+retrieve the list of `(key, value)` whose keys have a specific prefix.
+* Also, the list of the keys are provided in the lexicographic order.
+
+## The journaling.
```

Comment:
`## Journaling`

Let's also remove dots at the end of section titles.

Reply:
Ok.